akka-actor-typed/src/main/scala/akka/actor/typed/internal/TimerSchedulerImpl.scala

```diff
@@ -151,18 +149,38 @@ import scala.reflect.ClassTag
     // The scheduled TimerMsg is intercepted to guard against old messages enqueued
     // in mailbox before timer was canceled.
     // Intercept some signals to cancel timers when when restarting and stopping.
-    BehaviorImpl.intercept[T, TimerMsg](
-      beforeMessage = interceptTimerMsg,
-      beforeSignal = (ctx, sig) ⇒ {
-        sig match {
-          case PreRestart | PostStop ⇒ cancelAll()
-          case _                     ⇒ // unhandled
-        }
-        true
-      },
-      afterMessage = (ctx, msg, b) ⇒ b, // TODO optimize by using more ConstantFun
-      afterSignal = (ctx, sig, b) ⇒ b,
-      behavior)(ClassTag(classOf[TimerSchedulerImpl.TimerMsg]))
+    BehaviorImpl.intercept(new TimerInterceptor(this))(behavior).asInstanceOf[Behavior[T]]
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private final class TimerInterceptor[T](timerSchedulerImpl: TimerSchedulerImpl[T]) extends BehaviorInterceptor[AnyRef, T] {
+  import TimerSchedulerImpl._
+
+  override def aroundReceive(ctx: typed.ActorContext[AnyRef], msg: AnyRef, target: ReceiveTarget[T]): Behavior[T] = {
+    val intercepted = msg match {
+      case msg: TimerMsg ⇒ timerSchedulerImpl.interceptTimerMsg(ctx.asInstanceOf[ActorContext[TimerMsg]], msg)
+      case msg           ⇒ msg.asInstanceOf[T]
+    }
+
+    // filtering not possible
```

Comment:
?

Reply:
Ups. It was not possible but I solved that, and left the comment.